File: typescript/src/session-lakebase.ts
/**
 * LakebaseSessionStore — Postgres-backed session persistence over Databricks Lakebase.
 *
 * Lakebase is Databricks' managed Postgres. Compared to `DeltaSessionStore`:
 *
 *   - Lower per-operation latency (Postgres point lookups vs. SQL warehouse
 *     round-trips).
 *   - Better for high-frequency turns (chat UIs, interactive agents).
 *   - Worse for cheap durability across long-idle sessions.
 *
 * The store takes injectable Postgres-style executors (query + exec). The
 * caller wires up the actual transport — `pg`, `postgres`, a Databricks
 * Postgres proxy, whatever — typically with a `do_connect`-style hook that
 * mints fresh OAuth tokens via `ws.postgres.generate_database_credential()`.
 * The store stays narrow: SQL only.
 *
 * Schema (created on first use when `autoCreate=true`):
 *
 * ```sql
 * CREATE TABLE IF NOT EXISTS {table_name} (
 *   session_id  TEXT PRIMARY KEY,
 *   history     TEXT NOT NULL,    -- JSON-encoded list of message dicts
 *   state       TEXT NOT NULL,    -- JSON-encoded dict
 *   created_at  DOUBLE PRECISION,
 *   updated_at  DOUBLE PRECISION
 * )
 * ```
 */

import { z } from 'zod';
import type { Message } from './workflows/types.js';
import type { SessionSnapshot, SessionStore } from './workflows/session.js';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Row-returning executor — runs a parameterized SELECT and returns rows. */
export type PgQueryExecutor = (
  sql: string,
  params: Record<string, unknown>,
) => Promise<Array<Record<string, unknown>>>;

/** Side-effecting executor — runs parameterized INSERT/UPDATE/DELETE/DDL. */
export type PgExecExecutor = (
  sql: string,
  params: Record<string, unknown>,
) => Promise<void>;

const optionsSchema = z.object({
  tableName: z.string().min(1).optional(),
  autoCreate: z.boolean().optional(),
});

export interface LakebaseSessionStoreOptions {
  /** Row-returning executor for parameterized SELECT. */
  querySql: PgQueryExecutor;
  /** Side-effecting executor for parameterized DDL / DML. */
  execSql: PgExecExecutor;
  /** Table name (may include schema prefix). Defaults to `"apx_sessions"`. */
  tableName?: string;
  /** Run `CREATE TABLE IF NOT EXISTS` on first use. Defaults to true. */
  autoCreate?: boolean;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function isoToEpoch(iso: string): number {
  if (!iso) return 0;
  const ms = Date.parse(iso);
  return Number.isFinite(ms) ? ms / 1000 : 0;
}

function epochToIso(seconds: number): string {
  if (!Number.isFinite(seconds) || seconds <= 0) {
    return new Date(0).toISOString();
  }
  return new Date(seconds * 1000).toISOString();
}

// ---------------------------------------------------------------------------
// LakebaseSessionStore
// ---------------------------------------------------------------------------

export class LakebaseSessionStore implements SessionStore {
  readonly tableName: string;
  private readonly querySql: PgQueryExecutor;
  private readonly execSql: PgExecExecutor;
  private autoCreate: boolean;
  private created = false;

  constructor(options: LakebaseSessionStoreOptions) {
    optionsSchema.parse({ tableName: options.tableName, autoCreate: options.autoCreate });
    this.tableName = options.tableName ?? 'apx_sessions';
    this.querySql = options.querySql;
    this.execSql = options.execSql;
    this.autoCreate = options.autoCreate ?? true;
  }

  // -- table creation -----------------------------------------------------

  private async ensureTable(): Promise<void> {
    if (this.created || !this.autoCreate) return;
    // Flip the flag before issuing the DDL so we don't loop on a busted
    // permission/path setup (mirrors Python).
    this.created = true;
    const ddl =
      `CREATE TABLE IF NOT EXISTS ${this.tableName} (` +
      `  session_id TEXT PRIMARY KEY,` +
      `  history TEXT NOT NULL,` +
      `  state TEXT NOT NULL,` +
      `  created_at DOUBLE PRECISION,` +
      `  updated_at DOUBLE PRECISION` +
      `)`;
    try {
      await this.execSql(ddl, {});
    } catch (e) {
      console.warn(
        `LakebaseSessionStore: CREATE TABLE IF NOT EXISTS ${this.tableName} failed: ${String(e)} — ` +
          `subsequent reads/writes may fail until the table exists.`,
      );
    }
  }

  // -- SessionStore protocol --------------------------------------------

  async load(id: string): Promise<SessionSnapshot | null> {
    await this.ensureTable();
    const sql =
      `SELECT session_id, history, state, created_at, updated_at ` +
      `FROM ${this.tableName} WHERE session_id = :sid LIMIT 1`;
    let rows: Array<Record<string, unknown>>;
    try {
      rows = await this.querySql(sql, { sid: id });
    } catch (e) {
      console.warn(`LakebaseSessionStore.load(${id}) failed: ${String(e)} — returning null.`);
      return null;
    }
    if (!rows || rows.length === 0) return null;
    const row = rows[0]!;
    let messages: Message[];
    let state: Record<string, unknown>;
    try {
      messages = JSON.parse((row.history as string) || '[]') as Message[];
      state = JSON.parse((row.state as string) || '{}') as Record<string, unknown>;
    } catch (e) {
      console.warn(
        `LakebaseSessionStore.load(${id}): malformed JSON: ${String(e)} — returning empty session.`,
      );
      messages = [];
      state = {};
    }
    return {
      id: String(row.session_id ?? id),
      messages,
      state,
      createdAt: epochToIso(Number(row.created_at ?? 0)),
      updatedAt: epochToIso(Number(row.updated_at ?? 0)),
    };
  }

  async save(id: string, data: SessionSnapshot): Promise<void> {
    await this.ensureTable();
    const sql =
      `INSERT INTO ${this.tableName} ` +
      `(session_id, history, state, created_at, updated_at) ` +
      `VALUES (:sid, :history, :state, :created_at, :updated_at) ` +
      `ON CONFLICT (session_id) DO UPDATE SET ` +
      `  history = EXCLUDED.history, ` +
      `  state = EXCLUDED.state, ` +
      `  updated_at = EXCLUDED.updated_at`;
    await this.execSql(sql, {
      sid: id,
      history: JSON.stringify(data.messages ?? []),
      state: JSON.stringify(data.state ?? {}),
      created_at: isoToEpoch(data.createdAt),
      updated_at: isoToEpoch(data.updatedAt),
    });
  }

  async delete(id: string): Promise<void> {
    await this.ensureTable();
    const sql = `DELETE FROM ${this.tableName} WHERE session_id = :sid`;
    try {
      await this.execSql(sql, { sid: id });
    } catch (e) {
      console.warn(`LakebaseSessionStore.delete(${id}) failed: ${String(e)}`);
    }
  }

  async list(): Promise<string[]> {
    await this.ensureTable();
    const sql = `SELECT session_id FROM ${this.tableName}`;
    try {
      const rows = await this.querySql(sql, {});
      return rows.map((r) => String(r.session_id));
    } catch (e) {
      console.warn(`LakebaseSessionStore.list() failed: ${String(e)} — returning empty list.`);
      return [];
    }
  }

  // -- internal test hooks ------------------------------------------------

  /** @internal — Test helper to disable auto-create (mirrors Python). */
  _disableAutoCreate(): void {
    this.autoCreate = false;
    this.created = true;
  }
}

File: python/src/apx_agent/_session_lakebase.py
"""LakebaseSessionStore — Postgres-backed session persistence over Databricks Lakebase.

Lakebase is Databricks' managed Postgres. Compared to ``DeltaSessionStore``:

  * Lower per-operation latency (Postgres point lookups vs. SQL warehouse
    round-trips).
  * Better for high-frequency turns (chat UIs, interactive agents).
  * Worse for cheap durability across long-idle sessions (Delta scales
    storage independently; Lakebase charges for the running instance).

Both stores satisfy the same ``SessionStore`` protocol. Pick by workload:
chat-style → Lakebase; analytics-style multi-step pipelines → Delta.

This store takes a SQLAlchemy ``Engine`` rather than embedding the
Lakebase OAuth token rotation logic. The caller wires up the engine with
whatever auth pattern they prefer (typically a ``do_connect`` event
listener that calls ``ws.postgres.generate_database_credential()`` to
mint a fresh token). The store stays narrow: SQL only.

Schema (created on first put when ``auto_create=True``)::

    CREATE TABLE IF NOT EXISTS {table_name} (
        session_id  TEXT PRIMARY KEY,
        history     TEXT NOT NULL,   -- JSON-encoded list of message dicts
        state       TEXT NOT NULL,   -- JSON-encoded dict
        created_at  DOUBLE PRECISION,
        updated_at  DOUBLE PRECISION
    )

Requires the ``lakebase`` extra::

    pip install 'apx-agent[lakebase]'
"""

from __future__ import annotations

import json
import logging
import time
from typing import TYPE_CHECKING, Any

from ._session import Session, StoreError

if TYPE_CHECKING:
    from sqlalchemy.engine import Engine

logger = logging.getLogger(__name__)


_SQLALCHEMY_MISSING = (
    "LakebaseSessionStore requires SQLAlchemy. "
    "Install with: pip install 'apx-agent[lakebase]'"
)


def _require_sqlalchemy() -> Any:
    try:
        import sqlalchemy
    except ImportError as e:  # pragma: no cover — exercised only without extra
        raise ImportError(_SQLALCHEMY_MISSING) from e
    return sqlalchemy


class LakebaseSessionStore:
    """``SessionStore`` backed by a Lakebase (or any Postgres) instance.

    Args:
        engine: SQLAlchemy ``Engine`` connected to the Lakebase instance.
            The caller is responsible for token rotation — typically a
            ``do_connect`` event listener that mints fresh OAuth tokens
            via ``ws.postgres.generate_database_credential()``.
        table_name: Table name for session rows. Defaults to
            ``"apx_sessions"``. Quoted as a SQL identifier in queries,
            so it can include a schema prefix (e.g. ``"public.apx_sessions"``).
        auto_create: When True (default), runs ``CREATE TABLE IF NOT EXISTS``
            on first use. Set False when the table is provisioned by
            platform tooling.

    Example::

        from sqlalchemy import create_engine, event
        from databricks.sdk import WorkspaceClient
        from apx_agent import LakebaseSessionStore

        ws = WorkspaceClient()

        def _build_url() -> str:
            return "postgresql+psycopg://user@lakebase-host:5432/agentdb"

        engine = create_engine(_build_url(), pool_pre_ping=True, pool_recycle=1800)

        @event.listens_for(engine, "do_connect")
        def add_oauth_token(_dialect, _record, _args, kwargs):
            cred = ws.database.generate_database_credential(
                instance_names=["my-lakebase-instance"],
                request_id="apx-sessions",
            )
            kwargs["password"] = cred.token

        store = LakebaseSessionStore(engine=engine)
    """

    def __init__(
        self,
        *,
        engine: "Engine",
        table_name: str = "apx_sessions",
        auto_create: bool = True,
    ) -> None:
        _require_sqlalchemy()  # surfaces a friendly ImportError early
        self.engine = engine
        self.table_name = table_name
        self._auto_create = auto_create
        self._created = False

    # -- helpers -----------------------------------------------------------

    def _ensure_table(self) -> None:
        if self._created or not self._auto_create:
            return
        sa = _require_sqlalchemy()
        ddl = (
            f"CREATE TABLE IF NOT EXISTS {self.table_name} ("
            f"  session_id TEXT PRIMARY KEY,"
            f"  history TEXT NOT NULL,"
            f"  state TEXT NOT NULL,"
            f"  created_at DOUBLE PRECISION,"
            f"  updated_at DOUBLE PRECISION"
            f")"
        )
        try:
            with self.engine.begin() as conn:
                conn.execute(sa.text(ddl))
        except Exception as e:
            logger.warning(
                "LakebaseSessionStore: CREATE TABLE IF NOT EXISTS %s failed: %s — "
                "subsequent reads/writes may fail until the table exists.",
                self.table_name, e,
            )
        self._created = True

    # -- SessionStore protocol --------------------------------------------

    def get(self, session_id: str) -> Session | None:
        self._ensure_table()
        sa = _require_sqlalchemy()
        sql = sa.text(
            f"SELECT session_id, history, state, created_at, updated_at "
            f"FROM {self.table_name} WHERE session_id = :sid LIMIT 1"
        )
        try:
            with self.engine.connect() as conn:
                row = conn.execute(sql, {"sid": session_id}).mappings().first()
        except Exception as e:
            logger.warning(
                "LakebaseSessionStore.get(%s) failed: %s — raising StoreError "
                "(read failure is not a missing session).",
                session_id, e,
            )
            raise StoreError(
                f"LakebaseSessionStore.get({session_id!r}) failed: {e}"
            ) from e
        if row is None:
            return None
        try:
            history = json.loads(row["history"] or "[]")
            state = json.loads(row["state"] or "{}")
        except json.JSONDecodeError as e:
            logger.warning(
                "LakebaseSessionStore.get(%s): malformed JSON: %s — "
                "returning empty session.",
                session_id, e,
            )
            history, state = [], {}
        return Session(
            session_id=row["session_id"],
            history=history,
            state=state,
            created_at=float(row.get("created_at") or 0.0),
            updated_at=float(row.get("updated_at") or 0.0),
        )

    def put(self, session: Session) -> None:
        self._ensure_table()
        sa = _require_sqlalchemy()
        session.updated_at = time.time()
        # Postgres UPSERT via ON CONFLICT. The table_name is a controlled
        # identifier (set at construction), not user input, so it's safe
        # to inline. session_id and JSON payloads go through bind params.
        sql = sa.text(
            f"INSERT INTO {self.table_name} "
            f"(session_id, history, state, created_at, updated_at) "
            f"VALUES (:sid, :history, :state, :created_at, :updated_at) "
            f"ON CONFLICT (session_id) DO UPDATE SET "
            f"  history = EXCLUDED.history, "
            f"  state = EXCLUDED.state, "
            f"  updated_at = EXCLUDED.updated_at"
        )
        params = {
            "sid": session.session_id,
            "history": json.dumps(session.history, default=str),
            "state": json.dumps(session.state, default=str),
            "created_at": session.created_at,
            "updated_at": session.updated_at,
        }
        with self.engine.begin() as conn:
            conn.execute(sql, params)

    def delete(self, session_id: str) -> None:
        self._ensure_table()
        sa = _require_sqlalchemy()
        sql = sa.text(f"DELETE FROM {self.table_name} WHERE session_id = :sid")
        try:
            with self.engine.begin() as conn:
                conn.execute(sql, {"sid": session_id})
        except Exception as e:
            logger.warning(
                "LakebaseSessionStore.delete(%s) failed: %s",
                session_id, e,
            )
